### ISSUE TITLE
[MASSEMBLY-393] Cannot Override dependencyManagement

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#566]
**[deckrider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deckrider)** opened **[MASSEMBLY-393](https://issues.apache.org/jira/browse/MASSEMBLY-393?redirect=false)** and commented

This was not a problem in 2.2-beta-2.

I've provided a test case.  It should produce a zip file containing log4j-1.2.14.jar but instead contains log4j-1.2.12.jar

To run the test case:

unzip my-app.zip
cd my-app.zip
mvn clean package

This is critical for us, because we have a master parent pom where we keep all our versions in a dependencyManagement section, but on occasion we want to override them.  Upgrading from 2.2-beta-2 to 2.2-beta-3 broke that ability.

---

**Affects:** 2.2-beta-3

**Attachments:**
- [my-app.zip](https://issues.apache.org/jira/secure/attachment/12714109/my-app.zip) (_1.72 kB_)
- [my-app-pom-packaging.zip](https://issues.apache.org/jira/secure/attachment/12714309/my-app-pom-packaging.zip) (_1.72 kB_)

**Issue Links:**
- [MASSEMBLY-247](https://issues.apache.org/jira/browse/MASSEMBLY-247) versions of included dependencies in multi-module projects are not deterministic
 (_**"is duplicated by"**_)
- [MASSEMBLY-299](https://issues.apache.org/jira/browse/MASSEMBLY-299) assembly does not honnor \<dependencyManagement> directives
 (_**"is duplicated by"**_)
- [MASSEMBLY-432](https://issues.apache.org/jira/browse/MASSEMBLY-432) assembly misapplies depMgt and selects the wrong dependency for an archive
 (_**"is duplicated by"**_)
- [MASSEMBLY-235](https://issues.apache.org/jira/browse/MASSEMBLY-235) dependencySet ignores dependency management

7 votes, 7 watchers

**Comment by jira-importer** [comment]
**[deckrider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deckrider)** commented

FYI, using the attached example, the maven dependency plugin gets it right:

```
mvn org.apache.maven.plugins:maven-dependency-plugin:2.1:tree

[INFO] ------------------------------------------------------------------------
[INFO] Building my-app
[INFO]    task-segment: [org.apache.maven.plugins:maven-dependency-plugin:2.1:tree]
[INFO] ------------------------------------------------------------------------
[INFO] [dependency:tree]
[INFO] com.mycompany.app:my-app:jar:1.0-SNAPSHOT
[INFO] \- log4j:log4j:jar:1.2.14:compile
[INFO] ------------------------------------------------------------------------
```

And yet it still shows log4j-1.2.12.jar in the zip file:

```
unzip -l target/my-app-1.0-SNAPSHOT.zip
Archive:  target/my-app-1.0-SNAPSHOT.zip
  Length     Date   Time    Name
 --------    ----   ----    ----
        0  02-25-09 07:00   my-app-1.0-SNAPSHOT/
        0  02-25-09 07:00   my-app-1.0-SNAPSHOT/foo/
   358085  02-25-09 07:00   my-app-1.0-SNAPSHOT/foo/log4j-1.2.12.jar
     1701  02-25-09 07:00   my-app-1.0-SNAPSHOT/foo/my-app-1.0-SNAPSHOT.jar
 --------                   -------
   359786                   4 files
```

**Comment by jira-importer** [comment]
**[deckrider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deckrider)** commented

I see I forgot to change ...

\<packaging>jar\</packaging>

... to ...

\<packaging>pom\</packaging>

... in the test case, but after doing so, it doesn't alter the reported issue.

**Comment by jira-importer** [comment]
**[deckrider](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=deckrider)** commented

Uploaded an altered case which uses pom packaging instead of jar packaging.

**Comment by jira-importer** [comment]
**[Nathan Coast](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=nathancoast)** commented

confirmed, exact same issue - dependency plugin reports correct overridden version.  assembly plugin uses incorrect version from parent dependency management

**Comment by jira-importer** [comment]
**[Brian Lalor](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=blalor)** commented

It appears that 2.2-beta-3 reintroduces MASSEMBLY-235, which was fixed in 2.2-beta-2.

**Comment by jira-importer** [comment]
**[Chetan Narsude](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chetan)** commented

I am also facing a similar issue with 2.2-beta-5. For now I am using the workaround where I am re-specifying the dependencies in a dependencyManagement section of the child pom

From the code in manageArtifact in DefaultDependencyResolver

```java
 // Copied from DefaultArtifactCollector, SVN: http://svn.apache.org/repos/asf/maven/components/branches/maven-2.1.x@679206
    // with modifications to work with non-transitive resolution processes. 
    protected void manageArtifact( Artifact targetArtifact, Map managedVersions )
    {
        Artifact artifact = (Artifact) managedVersions.get( targetArtifact.getDependencyConflictId() );
        
        if ( artifact == null )
        {
            return;
        }

        // Before we update the version of the artifact, we need to know
        // whether we are working on a transitive dependency or not.  This
        // allows depMgmt to always override transitive dependencies, while
        // explicit child override depMgmt (viz. depMgmt should only
        // provide defaults to children, but should override transitives).
        // We can do this by calling isChildOfRootNode on the current node.

        if ( artifact.getVersion() != null )
        {
            if ( getLogger().isDebugEnabled() )
            {
                getLogger().debug( "Managing version for: " + targetArtifact.getDependencyConflictId() + " to: " + artifact.getVersion() );
            }
            
            targetArtifact.setVersion( artifact.getVersion() );
        }

        if ( artifact.getScope() != null )
        {
            if ( getLogger().isDebugEnabled() )
            {
                getLogger().debug( "Managing scope for: " + targetArtifact.getDependencyConflictId() + " to: " + artifact.getScope() );
            }
            
            targetArtifact.setScope( artifact.getScope() );
        }
    }
```

The comment indicates that intention is correct but the code says otherwise.

And looking at [DefaultArtifactCollector](http://svn.apache.org/repos/asf/maven/maven-2/branches/maven-2.2.x/maven-artifact/src/main/java/org/apache/maven/artifact/resolver/DefaultArtifactCollector.java) it looks like it is handled there
